persist/fs/index_lookup.go

```diff
@@ -0,0 +1,195 @@
+			// the file on disk was corrupted
+			if err != nil {
+				return -1, err
+			}
+			return indexOffset, nil
+		}
+
+		// idBytes is smaller than compBytes, go left
+		if comparison == -1 {
+			max = idx - 1
+			continue
+		}
+
+		// idBytes is larger than compBytes, go right
+		if comparison == 1 {
+			min = idx + 1
+			indexOffset, err := summaryBytesMetadata.IndexOffset(
+				il.summariesMmap, il.decoderStream, il.msgpackDecoder)
+			if err != nil {
+				return -1, err
+			}
+			// update the bestMatchSoFar everytime we move right
+			bestMatchSoFar = indexOffset
+			continue
+		}
+	}
+}
+
+func (il *nearestIndexOffsetLookup) close() error {
+	return munmap(il.summariesMmap)
+}
+
+// newNearestIndexOffsetLookupFromSummariesFile creates an nearestIndexOffsetLookup
+// from an index summaries file by reading the summaries file into an anonymous
+// mmap'd region, and also creating the slice of summaries offsets which is
+// required to binary search the data structure. It will also make sure that
+// the summaries file is sorted (which it always should be).
+func newNearestIndexOffsetLookupFromSummariesFile(
+	summariesFdWithDigest digest.FdWithDigestReader,
+	expectedSummariesDigest uint32,
+	decoder *xmsgpack.Decoder,
+	numEntries int,
+) (*nearestIndexOffsetLookup, error) {
+	summariesFd := summariesFdWithDigest.Fd()
+	stat, err := summariesFd.Stat()
+	if err != nil {
+		return nil, err
+	}
+	numBytes := stat.Size()
+
+	// Request an anonymous (non-file-backed) mmap region. Note that we're going
+	// to use the mmap'd region to store the read-only summaries data, but the mmap
+	// region itself needs to be writable so we can copy the bytes from disk
+	// into it
+	mmapResult, err := mmapBytes(numBytes, mmapOptions{read: true, write: true})
+	if err != nil {
+		return nil, err
+	}
+	summariesMmap := mmapResult.result
+
+	// Validate the bytes on disk using the digest, and read them into
+	// the mmap'd region
+	_, err = summariesFdWithDigest.ReadAllAndValidate(summariesMmap, expectedSummariesDigest)
+	if err != nil {
+		munmap(summariesMmap)
+		return nil, err
+	}
+
+	// Msgpack decode the entire summaries file (we need to store the offsets
+	// for the entries so we can binary-search it)
+	var (
+		decoderStream    = xmsgpack.NewDecoderStream(summariesMmap)
+		summariesOffsets = make([]xmsgpack.IndexSummaryToken, 0, numEntries)
+		lastReadID       []byte
+	)
+	decoder.Reset(decoderStream)
+
+	for read := 0; read < numEntries; read++ {
+		// We ignore the entry itself because we don't need any information from it
+		entry, entryMetadata, err := decoder.DecodeIndexSummary()
+		if err != nil {
+			munmap(summariesMmap)
+			return nil, err
+		}
+
+		// Make sure that all the IDs are sorted as we iterate, and return an error
+		// if they're not. This should never happen as files should be sorted on disk.
+		if lastReadID != nil && bytes.Compare(lastReadID, entry.ID) != -1 {
```

Comment:
`lastReadID` is not actually set at the bottom of this loop it seems?  

Can we add a test to catch this in the future too perhaps?

Reply:
Good catch, added the unit test